From ea52580945dc9bc8d0f63b8984e345ac65131ee1 Mon Sep 17 00:00:00 2001
From: Cosimo Cecchi <cosimoc@gnome.org>
Date: Mon, 17 Feb 2014 23:44:34 +0000
Subject: directory: collect all directories when invalidating

Otherwise we risk modifying the directories hash table from within the
iteration callback, which will trigger a critical warning.

https://bugzilla.gnome.org/show_bug.cgi?id=697890
---
diff --git a/libnemo-private/nemo-directory.c b/libnemo-private/nemo-directory.c
index 98dac79..8239cbe 100644
--- a/libnemo-private/nemo-directory.c
+++ b/libnemo-private/nemo-directory.c
@@ -244,28 +244,37 @@ nemo_directory_finalize (GObject *object)
 }
 
 static void
-invalidate_one_count (gpointer key, gpointer value, gpointer user_data)
+collect_all_directories (gpointer key, gpointer value, gpointer callback_data)
 {
 	NemoDirectory *directory;
-
-	g_assert (key != NULL);
-	g_assert (NEMO_IS_DIRECTORY (value));
-	g_assert (user_data == NULL);
+	GList **dirs;
 
 	directory = NEMO_DIRECTORY (value);
-	
-	nemo_directory_invalidate_count_and_mime_list (directory);
+	dirs = callback_data;
+
+	*dirs = g_list_prepend (*dirs, nemo_directory_ref (directory));
 }
 
 static void
 filtering_changed_callback (gpointer callback_data)
 {
+	GList *dirs, *l;
+	NemoDirectory *directory;
+
 	g_assert (callback_data == NULL);
 
+	dirs = NULL;
+	g_hash_table_foreach (directories, collect_all_directories, &dirs);
+
 	/* Preference about which items to show has changed, so we
 	 * can't trust any of our precomputed directory counts.
 	 */
-	g_hash_table_foreach (directories, invalidate_one_count, NULL);
+	for (l = dirs; l != NULL; l = l->next) {
+		directory = NEMO_DIRECTORY (l->data);
+		nemo_directory_invalidate_count_and_mime_list (directory);
+	}
+
+	nemo_directory_list_unref (dirs);
 }
 
 void
@@ -284,18 +293,6 @@ emit_change_signals_for_all_files (NemoDirectory *directory)
 	nemo_file_list_free (files);
 }
 
-static void
-collect_all_directories (gpointer key, gpointer value, gpointer callback_data)
-{
-	NemoDirectory *directory;
-	GList **dirs;
-
-	directory = NEMO_DIRECTORY (value);
-	dirs = callback_data;
-
-	*dirs = g_list_prepend (*dirs, nemo_directory_ref (directory));
-}
-
 void
 emit_change_signals_for_all_files_in_all_directories (void)
 {
--
cgit v0.9.2
